### PR TITLE
Remove calculate output shape from java side

### PR DESCRIPTION
**Pull request by rnett**
Mostly fixes #7901.

A few ops have issues that prevent removing the method now (#8131, #8129, #8128).

"Fake" Java-only ops have a `calculateOutputShape` that throws a `UnsupportedOperationException`.  Doing this in a method needed in each op isn't ideal, if there's an easy way to detect "fake" ops that would be easier, but I'm not aware of one.

### CODE REVIEW

**Comment by rnett** [comment]
Also note there is a test for cast on an empty array that NPEs.  I think its due to not having offline shape inference rather than a cast issue, but its worth taking a look at.

https://github.com/SkymindIO/deeplearning4j/blob/703de772ee4fd3f01eafbb51c0ea20415f103360/nd4j/nd4j-backends/nd4j-tests/src/test/java/org/nd4j/autodiff/samediff/SameDiffTests.java#L3557-L3566

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/samediff/SDVariable.java [hunk 1]
```diff
@@ -112,6 +112,14 @@ public boolean isConstant(){
...
     }
 
+    /**
+     * Returns true if the variable is empty.
+     *
+     * See {@link INDArray#isEmpty()}.
+     */
+    public boolean isEmpty(){
+        return getShapeDescriptor().isEmpty();
+    }
```
Comment (AlexDBlack) [inline nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/samediff/SDVariable.java]: I don't like this method for 2 reasons.
1. The shape descriptor isn't always available
2. It's not a constant thing - some operations (strided slice, for example) could return empty for some network inputs, but be non-empty for other inputs.

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/controlflow/compat/Enter.java [hunk 2]
```diff
@@ -70,11 +70,7 @@ public String opName() {
 
     @Override
     public List<LongShapeDescriptor> calculateOutputShape() {
-        if(arg().getArr() != null) {
-            return Collections.singletonList(LongShapeDescriptor.fromShape(arg().getShape(), arg().getArr().dataType()));
-        }
-        else
-            return Collections.emptyList();
+        throw new UnsupportedOperationException("Not an actual op, calculateOutputShape() is unneeded/unused.");
```
Comment (AlexDBlack) [inline nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/controlflow/compat/Enter.java]: I get the sentiment, but I don't like the "Not an actual op".
Technically it's true in terms of how we implement execution, but it's still an operation from the user point of view.
Better: "calculateOutputShape() is not supported for control flow operations".
Applies elsewhere as well, obviously.

Also: Let's define calculateOutputShape() for all of these once in BaseCompatOp instead.

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/transforms/BaseDynamicTransformOp.java [hunk 3]
```diff
@@ -47,66 +47,66 @@ public BaseDynamicTransformOp(INDArray[] inputs, INDArray[] outputs) {
...
+//                throw new RuntimeException("calculateOutputShape() failed for [" + this.opName() + "]", e);
+//            }
+//        } else {
+//            //Handle broadcast shape: [1,4]+[3,1] = [3,4]
+//            Shape.assertBroadcastable(firstArgShape, secondArgShape, this.getClass());
+//            val outShape = Shape.broadcastOutputShape(firstArgShape, secondArgShape);
+//
+//            return Collections.singletonList(LongShapeDescriptor.fromShape(outShape, dtypeZ));
+//        }
+//    }
```
Comment (AlexDBlack) [inline nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/transforms/BaseDynamicTransformOp.java]: Remove commented out code
